From ac6bfed90a9a2fc055d877b4e3cd0431e5170896 Mon Sep 17 00:00:00 2001
From: Arun Ravindran <arun.ravindran@intel.com>
Date: Mon, 19 Nov 2012 15:03:54 +0200
Subject: [PORT FROM MAIN] Telephony: Fix for com.android.phone has stopped during MTBF

BZ: 16858

This patch handles the exception ConcurrentModificationException
found during MTBF test.

Change-Id: Ia6db5a7de0afd4034fb8c7d5d8f81b265928e9ab
Orig-Change-Id: Ica2e7a52c03c9f2c561ed06807afdce439565e68
Category: aosp improvement
Domain: Telephony-Telephony-FWK
Origin: internal
Upstream-Candidate: yes
Signed-off-by: Arun Ravindran <arun.ravindran@intel.com>
---
 .../android/internal/telephony/CallManager.java    |   54 +++++++++++---------
 1 files changed, 29 insertions(+), 25 deletions(-)

diff --git a/src/java/com/android/internal/telephony/CallManager.java b/src/java/com/android/internal/telephony/CallManager.java
index be3aa78..69a7b90 100644
--- a/src/java/com/android/internal/telephony/CallManager.java
+++ b/src/java/com/android/internal/telephony/CallManager.java
@@ -31,6 +31,7 @@ import android.telephony.Rlog;
 
 import java.util.ArrayList;
 import java.util.Collections;
+import java.util.ConcurrentModificationException;
 import java.util.List;
 
 
@@ -1842,32 +1843,35 @@ public final class CallManager {
     public String toString() {
         Call call;
         StringBuilder b = new StringBuilder();
-
-        b.append("CallManager {");
-        b.append("\nstate = " + getState());
-        call = getActiveFgCall();
-        b.append("\n- Foreground: " + getActiveFgCallState());
-        b.append(" from " + call.getPhone());
-        b.append("\n  Conn: ").append(getFgCallConnections());
-        call = getFirstActiveBgCall();
-        b.append("\n- Background: " + call.getState());
-        b.append(" from " + call.getPhone());
-        b.append("\n  Conn: ").append(getBgCallConnections());
-        call = getFirstActiveRingingCall();
-        b.append("\n- Ringing: " +call.getState());
-        b.append(" from " + call.getPhone());
-
-        for (Phone phone : getAllPhones()) {
-            if (phone != null) {
-                b.append("\nPhone: " + phone + ", name = " + phone.getPhoneName()
-                        + ", state = " + phone.getState());
-                call = phone.getForegroundCall();
-                b.append("\n- Foreground: ").append(call);
-                call = phone.getBackgroundCall();
-                b.append(" Background: ").append(call);
-                call = phone.getRingingCall();
-                b.append(" Ringing: ").append(call);
+        try {
+            b.append("CallManager {");
+            b.append("\nstate = " + getState());
+            call = getActiveFgCall();
+            b.append("\n- Foreground: " + getActiveFgCallState());
+            b.append(" from " + call.getPhone());
+            b.append("\n  Conn: ").append(getFgCallConnections());
+            call = getFirstActiveBgCall();
+            b.append("\n- Background: " + call.getState());
+            b.append(" from " + call.getPhone());
+            b.append("\n  Conn: ").append(getBgCallConnections());
+            call = getFirstActiveRingingCall();
+            b.append("\n- Ringing: " +call.getState());
+            b.append(" from " + call.getPhone());
+
+            for (Phone phone : getAllPhones()) {
+                if (phone != null) {
+                    b.append("\nPhone: " + phone + ", name = " + phone.getPhoneName()
+                            + ", state = " + phone.getState());
+                    call = phone.getForegroundCall();
+                    b.append("\n- Foreground: ").append(call);
+                    call = phone.getBackgroundCall();
+                    b.append(" Background: ").append(call);
+                    call = phone.getRingingCall();
+                    b.append(" Ringing: ").append(call);
+                }
             }
+        } catch (ConcurrentModificationException e) {
+            Rlog.e(LOG_TAG, "CallManager toString Exception:" + e);
         }
         b.append("\n}");
         return b.toString();
-- 
1.7.4.1

